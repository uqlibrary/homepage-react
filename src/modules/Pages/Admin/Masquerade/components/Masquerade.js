import React, { useState } from 'react';
import PropTypes from 'prop-types';

import Button from '@mui/material/Button';
import Grid from '@mui/material/Grid';
import { styled } from '@mui/material/styles';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';

import { StandardCard } from 'modules/SharedComponents/Toolbox/StandardCard';
import { StandardPage } from 'modules/SharedComponents/Toolbox/StandardPage';
import txt from './masquerade.locale';

const StyledStandardCard = styled(StandardCard)(() => ({
    backgroundColor: 'white',
}));

const Masquerade = ({ account }) => {
    const [userName, setUserName] = useState('');
    const [loading, setLoading] = useState(false);

    const masqueradeAs = event => {
        if ((event && event.key && event.key !== 'Enter') || userName.length === 0) return;

        setLoading(true);

        sessionStorage.removeItem('userAccount');

        let redirectUrl = `${window.location.protocol}//${window.location.hostname}`;
        const isDevBranch = window.location.hostname === 'homepage-development.library.uq.edu.au';
        redirectUrl += isDevBranch ? /* istanbul ignore next */ window.location.pathname : '/';
        window.location.assign(
            `https://auth.library.uq.edu.au/masquerade?user=${userName}&return=${window.btoa(redirectUrl)}`,
        );
    };

    const usernameChanged = event => {
        setUserName(event.target.value);
    };

    const userCapabilityStatement = canMasqueradeType => {
        return canMasqueradeType && canMasqueradeType === 'readonly'
            ? txt.user.access.readonly.capabilityStatement
            : txt.user.access.full.capabilityStatement;
    };

    return (
        <StandardPage title={txt.title}>
            <StyledStandardCard noHeader standardCardId="masquerade">
                <Typography>{userCapabilityStatement(account.canMasqueradeType)}</Typography>
                <Grid container spacing={3} alignItems={'flex-end'} style={{ marginTop: 12 }}>
                    <Grid item xs>
                        <TextField
                            variant="standard"
                            data-testid="masquerade-userName"
                            fullWidth
                            label={txt.labels.hint}
                            value={userName}
                            onChange={usernameChanged}
                            InputProps={{ onKeyPress: masqueradeAs }}
                        />
                    </Grid>
                    <Grid item xs={12} sm={'auto'}>
                        <Button
                            variant="contained"
                            id="submitMasquerade"
                            data-analyticsid="submitMasquerade"
                            data-testid="masquerade-submit"
                            fullWidth
                            color="primary"
                            children={txt.labels.submit}
                            disabled={loading}
                            onClick={masqueradeAs}
                            onKeyUp={masqueradeAs}
                        />
                    </Grid>
                </Grid>
            </StyledStandardCard>
        </StandardPage>
    );
};

Masquerade.propTypes = {
    account: PropTypes.object.isRequired,
};

export default React.memo(Masquerade);
