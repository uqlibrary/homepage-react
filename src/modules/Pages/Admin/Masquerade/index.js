export { default as Masquerade } from './containers/Masquerade';
