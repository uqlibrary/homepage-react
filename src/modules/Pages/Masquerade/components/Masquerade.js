import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import txt from './masqueradeLocale';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import { StandardCard } from 'modules/SharedComponents/Toolbox/StandardCard';
import { StandardPage } from 'modules/SharedComponents/Toolbox/StandardPage';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';

export default class Masquerade extends PureComponent {
    static propTypes = {
        account: PropTypes.object.isRequired,
    };

    constructor(props) {
        super(props);
        this.state = {
            userName: '',
            loading: false,
        };
    }

    _masqueradeAs = event => {
        if ((event && event.key && event.key !== 'Enter') || this.state.userName.length === 0) return;

        this.setState({
            loading: true,
        });

        const redirectUrl = `${window.location.protocol}//${window.location.hostname}/`;
        window.location.assign(
            `https://auth.library.uq.edu.au/masquerade?user=${this.state.userName}&return=${window.btoa(redirectUrl)}`,
        );
    };

    _usernameChanged = event => {
        this.setState({
            userName: event.target.value,
        });
    };

    userCapabilityStatement = canMasqueradeType => {
        return canMasqueradeType && canMasqueradeType === 'readonly'
            ? txt.user.access.readonly.capabilityStatement
            : txt.user.access.full.capabilityStatement;
    };

    render() {
        return (
            <StandardPage title={txt.title} goBackFunc={() => history.back()} goBackTooltip="Go back">
                <div className="layout-card">
                    <StandardCard noHeader>
                        <Grid container spacing={1}>
                            <Grid item xs={12} id="masquerade">
                                <Typography>
                                    {this.userCapabilityStatement(this.props.account.canMasqueradeType)}
                                </Typography>
                                <Grid container spacing={3} alignItems={'flex-end'} style={{ marginTop: 12 }}>
                                    <Grid item xs>
                                        <TextField
                                            data-testid="masquerade-userName"
                                            fullWidth
                                            id="userName"
                                            label={txt.labels.hint}
                                            onChange={this._usernameChanged}
                                            onKeyPress={this._masqueradeAs}
                                            value={this.state.userName}
                                        />
                                    </Grid>
                                    <Grid item xs={12} sm={'auto'}>
                                        <Button
                                            children={txt.labels.submit}
                                            color="primary"
                                            data-testid="masquerade-submit"
                                            disabled={this.state.loading}
                                            fullWidth
                                            id="submitMasquerade"
                                            onClick={this._masqueradeAs}
                                            variant="contained"
                                        />
                                    </Grid>
                                </Grid>
                            </Grid>
                        </Grid>
                    </StandardCard>
                </div>
            </StandardPage>
        );
    }
}
