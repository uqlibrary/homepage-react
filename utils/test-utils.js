/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

import preview from 'jest-preview';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';
import { createMemoryHistory } from 'history';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');

const { configure } = domTestingLib;

configure(config => ({
    ...config,
    testIdAttribute: 'id',
}));

const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <MuiPickersUtilsProvider utils={MomentUtils}>{props.children}</MuiPickersUtilsProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {},
) => {
    return render(
        <AllTheProviders>
            <Router history={history}>{ui}</Router>
        </AllTheProviders>,
    );
};

export const RenderWithRouter = ({
    children,
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
}) => (
    <AllTheProviders>
        <Router history={history}>{children}</Router>
    </AllTheProviders>
);
export const renderWithRedux = ({ initialState }) => render => {
    return {
        ...render,
        store: getStore({ initialState, history: render.history }),
    };
};

export const withRouter = ({
    route = '/',
    path = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
} = {}) => WrappedComponent => {
    return (
        <Router history={history}>
            <Route path={path} children={WrappedComponent} />
        </Router>
    );
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

// eslint-disable-next-line react/prop-types
export const WithRedux = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>{children}</Provider>
);

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    renderWithRouter,
    renderWithRedux,
    withRedux,
    withRouter,
    WithRedux,
    WithReduxStore,
    RenderWithRouter,
    preview,
};
