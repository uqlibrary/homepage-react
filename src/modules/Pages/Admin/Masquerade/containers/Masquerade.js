import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';
import * as actions from 'actions';

import Masquerade from '../components/Masquerade';

const mapStateToProps = state => {
    return {
        ...state.get('accountReducer'),
    };
};

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(actions, dispatch),
    };
}

let MasqueradeContainer = connect(mapStateToProps, mapDispatchToProps)(Masquerade);
MasqueradeContainer = withRouter(MasqueradeContainer);

export default MasqueradeContainer;
