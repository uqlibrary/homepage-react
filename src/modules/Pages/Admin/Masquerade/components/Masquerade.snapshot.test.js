import React from 'react';
import Masquerade from './Masquerade';
import { rtlRender } from 'test-utils';

function setup(testProps = {}) {
    const props = {
        author: testProps.author || null,
        actions: testProps.actions || {},
        history: testProps.history || {
            push: jest.fn(),
        },
        account: testProps.account || {},
        ...testProps,
    };
    return rtlRender(<Masquerade {...props} />);
}

describe('Component Masquerade', () => {
    it('Should render form as expected', () => {
        const { container } = setup();
        expect(container).toMatchSnapshot();
    });

    it('Should render readonly description', () => {
        const props = {
            account: { canMasqueradeType: 'readonly' },
        };
        const { container } = setup(props);
        expect(container).toMatchSnapshot();
    });
});
