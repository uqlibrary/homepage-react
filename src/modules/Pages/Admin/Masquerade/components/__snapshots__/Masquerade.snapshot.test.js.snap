// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Masquerade Should render form as expected 1`] = `
<div>
  <div
    class="layout-card"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1bzpx2o-MuiGrid-root"
      data-testid="StandardPage"
      id="StandardPage"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true makeStyles-title-4 css-1vd824g-MuiGrid-root"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4 makeStyles-layoutTitle-3 css-v0z470-MuiTypography-root"
          data-testid="StandardPage-title"
        >
          Masquerade
        </h1>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root Cards-card-6 StandardCard  css-bhp9pd-MuiPaper-root-MuiCard-root"
          data-analyticsid="standard-card"
          data-testid="standard-card"
          id="standard-card"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            data-testid="standard-card-content"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
                id="masquerade"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-u1nwny-MuiTypography-root"
                >
                  <span>
                    <strong>
                      WARNING!!
                    </strong>
                     When masquerading as a user, you will effectively become that user, and changes you make will apply to the account!
                  </span>
                </p>
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-1vne64g-MuiGrid-root"
                  style="margin-top: 12px;"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                      data-testid="masquerade-userName"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-yexcec-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for="userName"
                        id="userName-label"
                      >
                        Enter a UQ staff or student username (eg. uqjsmith1 or s123456)
                      </label>
                      <div
                        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-199lw75-MuiInputBase-root-MuiInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                          id="userName"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-auto css-pr9bjb-MuiGrid-root"
                  >
                    <button
                      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-pitokd-MuiButtonBase-root-MuiButton-root"
                      data-testid="masquerade-submit"
                      id="submitMasquerade"
                      tabindex="0"
                      type="button"
                    >
                      Masquerade
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Component Masquerade Should render readonly description 1`] = `
<div>
  <div
    class="layout-card"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1bzpx2o-MuiGrid-root"
      data-testid="StandardPage"
      id="StandardPage"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true makeStyles-title-14 css-1vd824g-MuiGrid-root"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4 makeStyles-layoutTitle-13 css-v0z470-MuiTypography-root"
          data-testid="StandardPage-title"
        >
          Masquerade
        </h1>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root Cards-card-16 StandardCard  css-bhp9pd-MuiPaper-root-MuiCard-root"
          data-analyticsid="standard-card"
          data-testid="standard-card"
          id="standard-card"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            data-testid="standard-card-content"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
                id="masquerade"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-u1nwny-MuiTypography-root"
                >
                  <span>
                    <strong>
                      NOTE:
                    </strong>
                     As a read-only masquerader, you can view all parts of the profile, but you are not able to make any changes to the account.
                  </span>
                </p>
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-1vne64g-MuiGrid-root"
                  style="margin-top: 12px;"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                      data-testid="masquerade-userName"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-yexcec-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for="userName"
                        id="userName-label"
                      >
                        Enter a UQ staff or student username (eg. uqjsmith1 or s123456)
                      </label>
                      <div
                        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-199lw75-MuiInputBase-root-MuiInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                          id="userName"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-auto css-pr9bjb-MuiGrid-root"
                  >
                    <button
                      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-pitokd-MuiButtonBase-root-MuiButton-root"
                      data-testid="masquerade-submit"
                      id="submitMasquerade"
                      tabindex="0"
                      type="button"
                    >
                      Masquerade
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
